src/kv/test/kv_test.cpp

```diff
@@ -281,18 +281,55 @@ TEST_CASE_TEMPLATE("foreach", MapImpl, RawMapTypes, ExperimentalMapTypes)
 
   SUBCASE("Early termination")
   {
-    kv::Tx tx;
-    auto view = tx.get_view(map);
-    view->put("key1", "value1");
-    view->put("key2", "value2");
-    view->put("key3", "value3");
-    size_t ctr = 0;
-    view->foreach([&ctr](const auto& key, const auto& value) {
-      ++ctr;
-      return ctr <= 1; // Continue after the first, but not the second (so never
-                       // see the third)
-    });
-    REQUIRE(ctr == 2);
+    {
+      kv::Tx tx;
+      auto view = tx.get_view(map);
+      view->put("key1", "value1");
+      view->put("key2", "value2");
+      view->put("key3", "value3");
+      size_t ctr = 0;
+      view->foreach([&ctr](const auto& key, const auto& value) {
+        ++ctr;
+        return ctr < 2; // Continue after the first, but not the second (so
+                        // never see the third)
+      });
+      REQUIRE(ctr == 2);
+      REQUIRE(tx.commit() == kv::CommitSuccess::OK);
+    }
+
+    {
+      kv::Tx tx;
+      auto view = tx.get_view(map);
+      view->put("key4", "value4");
+      view->put("key5", "value5");
+
+      {
+        size_t ctr = 0;
+        view->foreach([&ctr](const auto&, const auto&) {
+          ++ctr;
+          return ctr < 2; //< See only committed state
```

Comment:
I am not sure I understand this test. From looking at the next `foreach()`, it would continue and loop through all 5 keys?

Reply:
As discussed, this makes perfect sense now!

Reply:
Discussed offline, hopefully clarified here for anyone reading in future: This test would fail under the previous code. The lambda would be called twice for the committed state (returning false the second time), and then _again for the first write_. So the counter would be 3, and our lambda was called after it returned false.